File: plugins/claude-code.js
import { ProtocolPlugin } from './base.js';

/**
 * Claude Code 协议插件
 * 将 CherryStudio 等第三方客户端的请求伪装成 Claude Code CLI
 */
export class ClaudeCodePlugin extends ProtocolPlugin {
  name = 'ClaudeCode';
  version = '1.0.0';
  description = 'Transform requests to Claude Code CLI format';

  // Claude Code system 模板
  static SYSTEM_TEMPLATE = [
    {
      type: 'text',
      text: 'You are Claude Code, Anthropic\'s official CLI for Claude.',
      cache_control: {
        type: 'ephemeral'
      }
    }
  ];

  // Claude Code 请求头特征（基于官方 2.0.24 版本）
  static HEADERS = {
    'accept': 'application/json',
    'anthropic-beta': 'claude-code-20250219',
    'anthropic-dangerous-direct-browser-access': 'true',
    'anthropic-version': '2023-06-01',
    'content-type': 'application/json',
    'user-agent': 'claude-cli/2.0.24 (external, cli)',
    'x-app': 'cli',
    'x-stainless-arch': 'x64',
    'x-stainless-lang': 'js',
    'x-stainless-os': 'Linux',
    'x-stainless-runtime': 'node',
    'x-stainless-runtime-version': 'v22.16.0'
  };

  // Session 缓存
  cachedSession = null;
  sessionExpireTime = 0;
  SESSION_TTL = 12 * 60 * 60 * 1000; // 12小时

  /**
   * 检测是否为需要转换的请求
   */
  detect(body, headers) {
    const userAgent = headers['user-agent'] || '';

    // 特征1: User-Agent包含CherryStudio
    const hasCherryUA = userAgent.includes('CherryStudio');

    // 特征2: tools字段缺失或为空
    const hasNoTools = !body.tools || body.tools.length === 0;

    // 特征3: system不包含"You are Claude Code"
    const notClaudeCode = !body.system?.some(s =>
      s.text?.includes('You are Claude Code')
    );

    return hasCherryUA || (hasNoTools && notClaudeCode);
  }

  /**
   * 转换请求体
   */
  transform(body, headers) {
    // 1. 替换system字段为Claude Code格式
    body.system = ClaudeCodePlugin.SYSTEM_TEMPLATE;

    // 2. 添加metadata（优先复用，否则使用12小时轮换session）
    if (!body.metadata?.user_id) {
      body.metadata = this.generateMetadata();
    }

    // 3. 移除Cherry特有的thinking字段
    if (body.thinking) {
      delete body.thinking;
    }

    return body;
  }

  /**
   * 获取目标请求头
   */
  getHeaders(apiKey) {
    return {
      ...ClaudeCodePlugin.HEADERS,
      'authorization': `Bearer ${apiKey}`
    };
  }

  /**
   * 生成metadata（12小时轮换session）
   */
  generateMetadata() {
    const now = Date.now();

    // 如果session未过期，复用
    if (this.cachedSession && now < this.sessionExpireTime) {
      return this.cachedSession;
    }

    // 生成新session
    const timestamp = new Date().toISOString().split('T')[0];
    const halfDayMark = Math.floor(now / this.SESSION_TTL);
    const sessionId = `88code-cherry-${timestamp}-${halfDayMark}`;

    this.cachedSession = {
      user_id: `user_88code_proxy_account__session_${sessionId}`
    };
    this.sessionExpireTime = now + this.SESSION_TTL;

    return this.cachedSession;
  }
}

File: service-cc/adapters/claude_code.py
"""
Claude Code CLI 客户端适配器

检测官方 Claude Code CLI 并直接透传请求
使用统一的 common.adapters 接口
"""

from common.adapters import AdapterContext, ClientAdapter, TransformResult
from common.logger import get_logger

logger = get_logger(__name__)


class ClaudeCodeAdapter(ClientAdapter):
    """Claude Code CLI 客户端适配器"""

    name = "claude_code"
    priority = 100  # 最高优先级 (官方客户端)
    version = "1.0.0"

    def detect(self, ctx: AdapterContext) -> bool:
        """
        检测是否为 Claude Code CLI

        检测规则:
        1. User-Agent 包含 "claude-cli"
        2. 或者包含 "anthropic-beta: claude-code-"

        Args:
            ctx: 适配器上下文

        Returns:
            是否为 Claude Code CLI
        """
        # 检查 User-Agent
        user_agent = ctx.raw_headers.get("user-agent", "").lower()
        if "claude-cli" in user_agent:
            logger.info("detected_claude_code", source="user_agent", user_agent=user_agent)
            return True

        # 检查 anthropic-beta 头
        anthropic_beta = ctx.raw_headers.get("anthropic-beta", "").lower()
        if "claude-code-" in anthropic_beta:
            logger.info("detected_claude_code", source="anthropic_beta", beta=anthropic_beta)
            return True

        logger.debug("not_claude_code", user_agent=user_agent)
        return False

    def transform(self, ctx: AdapterContext) -> TransformResult:
        """
        转换 Claude Code CLI 请求

        Claude Code CLI 的请求已经是完整格式,直接透传即可。
        不做任何修改,保留所有字段 (system, tools, thinking, metadata 等)

        Args:
            ctx: 适配器上下文

        Returns:
            转换结果 (直接透传原始请求体)
        """
        body = ctx.raw_body  # 直接使用原始请求,不做任何修改

        # 记录请求信息 (仅用于调试)
        thinking_enabled = "thinking" in body
        has_metadata = "metadata" in body

        logger.info(
            "claude_code_passthrough",
            adapter=self.name,
            system_blocks=len(body.get("system", [])),
            tools_count=len(body.get("tools", [])),
            thinking_enabled=thinking_enabled,
            has_metadata=has_metadata,
            user_id=body.get("metadata", {}).get("user_id") if has_metadata else None,
        )

        return TransformResult(
            body=body,  # 完全透传
            extra_headers={},  # 不注入任何额外 headers
            metadata={
                "adapter": self.name,
                "version": self.version,
                "is_official_client": True,
                "passthrough": True,  # 标记为透传模式
            },
        )
